File: 프로그래머스/lv2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.swift
import Foundation

func solution(_ bridge_length:Int, _ weight:Int, _ truck_weights:[Int]) -> Int {
    var truck = truck_weights
    var CarsOnBridge : [Int] = []
    var CarsLocation : [Int] = []
    var count : Int = 0
    var BridgeWeight : Int = 0
    
    truck.reverse()
    
    while truck.count != 0 || CarsLocation.count != 0 {
        count += 1
        CarsLocation = CarsLocation.map({ $0 + 1 })
        
        if CarsLocation.contains(bridge_length + 1) {
            CarsLocation.reverse()
            CarsLocation.popLast()
            CarsLocation.reverse()
            
            CarsOnBridge.reverse()
            BridgeWeight -= CarsOnBridge.popLast()!
            CarsOnBridge.reverse()
//            print("차량 빠짐, 현재 시간 \(count)")
        }
        
        if !truck.isEmpty {
            if BridgeWeight + truck.last! <= weight {
                var turn : Int = truck.popLast()!
                CarsLocation.append(1)
                CarsOnBridge.append(turn)
                BridgeWeight += turn
//                print("차량 \(turn) 올라감, 현재 시간 \(count)")
            }
        }
    }
    return count
}

File: 프로그래머스/lv2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
def solution(bridge_length, weight, truck_weights):

    bridge_on = []
    truck_location = []
    time = 0
    
    while truck_weights or bridge_on :
        
        if truck_weights:
        
            if sum(bridge_on) + truck_weights[0] <= weight:
                bridge_on.append(truck_weights.pop(0)) # 선입선출
                truck_location.append(0)
            
        time += 1
        truck_location = list(map(lambda x:x + 1, truck_location))
        
        #print("시간, 차 현황, 위치", time, bridge_on, truck_location)
        
        if truck_location[0] == bridge_length:
            bridge_on.pop(0)
            truck_location.pop(0)
    
    return time + 1